nexus/src/app/sagas/instance_start.rs

```diff
@@ -0,0 +1,710 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implements a saga that starts an instance.
+
+use super::{NexusActionContext, NexusSaga, SagaInitError};
+use crate::{
+    app::{
+        instance::WriteBackUpdatedInstance,
+        sagas::{declare_saga_actions, retry_until_known_result},
+    },
+    authn, authz,
+    db::{self, identity::Resource, lookup::LookupPath},
+};
+use omicron_common::api::external::{Error, InstanceState};
+use serde::{Deserialize, Serialize};
+use sled_agent_client::types::InstanceStateRequested;
+use slog::info;
+use steno::ActionError;
+use uuid::Uuid;
+
+/// Parameters to the instance start saga.
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Params {
+    pub instance: db::model::Instance,
+
+    /// Authentication context to use to fetch the instance's current state from
+    /// the database.
+    pub serialized_authn: authn::saga::Serialized,
+
+    /// True if the saga should configure Dendrite and OPTE configuration for
+    /// this instance. This allows the instance create saga to do this work
+    /// prior to invoking the instance start saga as a subsaga without repeating
+    /// these steps.
+    pub ensure_network: bool,
+}
+
+declare_saga_actions! {
+    instance_start;
+
+    MARK_AS_STARTING -> "starting_state" {
+        + sis_move_to_starting
+        - sis_move_to_starting_undo
+    }
+
+    // This saga should eventually trigger a RPW that updates all this state
+    // (see PR #3804), but for now it manually ensures that the correct
+    // networking state has been propagated.
+    DPD_ENSURE -> "dpd_ensure" {
+        + sis_dpd_ensure
+        - sis_dpd_ensure_undo
+    }
+
+    V2P_ENSURE -> "v2p_ensure" {
+        + sis_v2p_ensure
+        - sis_v2p_ensure_undo
+    }
+
+    ENSURE_REGISTERED -> "ensure_registered" {
+        + sis_ensure_registered
+        - sis_ensure_registered_undo
+    }
+
+    ENSURE_RUNNING -> "ensure_running" {
+        + sis_ensure_running
+    }
+}
+
+#[derive(Debug)]
+pub struct SagaInstanceStart;
+impl NexusSaga for SagaInstanceStart {
+    const NAME: &'static str = "instance-start";
+    type Params = Params;
+
+    fn register_actions(registry: &mut super::ActionRegistry) {
+        instance_start_register_actions(registry);
+    }
+
+    fn make_saga_dag(
+        _params: &Self::Params,
+        mut builder: steno::DagBuilder,
+    ) -> Result<steno::Dag, SagaInitError> {
+        builder.append(mark_as_starting_action());
+        builder.append(dpd_ensure_action());
+        builder.append(v2p_ensure_action());
```

Comment:
Would both of these steps just become "trigger the Nexus RPW for NAT" once https://github.com/oxidecomputer/omicron/pull/3804 merges?

Reply:
Yep! There's a comment on line 47 about this; I can file a tracking issue and link up to that issue number (will leave this thread open until I do).